Add CrownJewel GRPC service and related messages

jakule opened the pull request:
This commit introduces the CrownJewel service, along with its corresponding request and response messages. The changes include new proto definitions for creating, getting, updating, and deleting CrownJewels. Functions for pagination are also introduced for the GetCrownJewels method. This new service will provide methods to manage critical resources considered 'CrownJewels' for the organization.

Implements messages described in the RFD https://github.com/gravitational/access-graph/pull/655

jakule committed:
Add CrownJewel GRPC service and related messages

This commit introduces the CrownJewel service, along with its corresponding request and response messages. The changes include new proto definitions for creating, getting, updating, and deleting CrownJewels. Functions for pagination are also introduced for the GetCrownJewels method. This new service will provide methods to manage critical resources considered 'CrownJewels' for the organization.

tigrato commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,76 @@
+  // Multiple kinds can be provided to match multiple kinds.
+  repeated string kinds = 2;
+  // Labels is a set of labels.
+  map<string, string> labels = 3;
```
Use the same approach as aws and just enforce it to be non-nil

jakule replied:
@tigrato I can, but in almost all places in our internal implementation, we're using maps to store our labels. Even our metadata struct uses it https://github.com/gravitational/teleport/blob/5bc69e06654ff4eea89b658326bfd2cf2bbb6aed/api/proto/teleport/header/v1/metadata.proto#L33
so I'm not sure why here we should diverge from it?

tigrato replied:
In reality we should allow:

`key:[]*string`  where the string works as an or operator. 

That's how we did all matchers in teleport because you might want to select env: [prod, super-prod] without having to duplicate the matchers.

My idea is to

repeated TagMatcher labels = 5;


message TAGMatcher {
    string Key = 1;
    repeated google.protobuf.StringValue values =2;
}

jakule committed:
Update TeleportMatcher labels to use TeleportLabel

This update changes the previous map data type of 'labels' in TeleportMatcher to use TeleportLabel, allowing for more flexibility. Also, the TeleportLabel message is created with 'key' and 'values' attributes in the api/proto/teleport directory. This new structure complements the changes made in crownjewel.pb.go.

tigrato commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,84 @@
+  // Key is the key of the tag.
+  string key = 1;
+  // Value is the value of the tag.
+  google.protobuf.StringValue value = 2;
```
should we also make this field repeated?

tigrato commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,84 @@
+  // Key is the key of the label.
+  string key = 1;
+  // Values is a list of label values to match.
+  repeated google.protobuf.StringValue values = 2;
```
if we are using a separate representation between teleport and aws, we can just use repeated string here.
My suggestion was if we reused the same label matcher for aws and teleport

tigrato commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,84 @@
+}
+
+// TeleportLabel is a label that is attached to a Teleport resource.
+message TeleportLabel {
```
```suggestion
message TeleportLabelMatcher {
```

jakule replied:
If we call it label matcher, then every field in the TeleportMatcher should also be called matcher... That a lot of matches 😅

jakule committed:
Update label and tag value types in crownjewel.proto and crownjewel.pb.go

This commit changes the value types of labels and tags in crownjewel.proto and crownjewel.pb.go. Instead of using google.protobuf.StringValue, the label now uses simple strings for values, and a list of google.protobuf.StringValue for the tag values.

jakule commented on api/proto/teleport/crownjewel/v1/crownjewel_service.proto:
```diff
@@ -0,0 +1,83 @@
+  // DeleteCrownJewel deletes a CrownJewel.
+  rpc DeleteCrownJewel(DeleteCrownJewelRequest) returns (google.protobuf.Empty);
+  // DeleteAllCrownJewels deletes all CrownJewels.
+  rpc DeleteAllCrownJewels(DeleteAllCrownJewelsRequest) returns (google.protobuf.Empty);
```
Nope, removed

jakule commented on api/proto/teleport/crownjewel/v1/crownjewel_service.proto:
```diff
@@ -0,0 +1,83 @@
+// CrownJewelService is a service that provides methods to manage CrownJewels.
+service CrownJewelService {
+  // CreateCrownJewel creates a new CrownJewel.
+  rpc CreateCrownJewel(CreateCrownJewelRequest) returns (CreateCrownJewelResponse);
```
Changed

jakule commented on api/proto/teleport/crownjewel/v1/crownjewel_service.proto:
```diff
@@ -0,0 +1,83 @@
+  // CreateCrownJewel creates a new CrownJewel.
+  rpc CreateCrownJewel(CreateCrownJewelRequest) returns (CreateCrownJewelResponse);
+  // GetCrownJewels returns a list of CrownJewels. It supports pagination.
+  rpc GetCrownJewels(GetCrownJewelsRequest) returns (GetCrownJewelsResponse);
```
Changed

jakule committed:
RFD 153

jakule committed:
Rename CrownJewels proto messages

Updated protobuf message naming for consistency and clarity. The GetCrownJewelsRequest and GetCrownJewelsResponse were renamed to ListCrownJewelsRequest and ListCrownJewelsResponse respectively. This change better reflects their functionality.

jakule committed:
Rename CrownJewels proto messages

Updated protobuf message naming for consistency and clarity. The GetCrownJewelsRequest and GetCrownJewelsResponse were renamed to ListCrownJewelsRequest and ListCrownJewelsResponse respectively. This change better reflects their functionality.

smallinsky commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,90 @@
+  // Multiple kinds can be provided to match multiple kinds.
+  repeated string kinds = 2;
+  // Labels is a set of labels.
+  repeated TeleportLabel labels = 3;
```
nit:  Can we reuse [teleprot.labelv.1Lables](https://github.com/gravitational/teleport/blob/9d7c59f9a40f6fa990106b1976ecb5925d1e58af/api/proto/teleport/label/v1/label.proto#L22) ? 
```suggestion
  repeated teleport.label.v1.Label labels = 3;
```

smallinsky commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,90 @@
+
+// TeleportMatcher represents a matcher for Teleport resources.
+message TeleportMatcher {
+  // Name is the name of the resource. When the name is provided, it will match
+  // the resource with the same name.
+  string name = 1;
```
nit: 

Do we want to support array of names  like in case of kinds ?

jakule replied:
I'd say no. You can specify multiple matches. That allows to provide multiple names. Having multiple names and multiple matches sounds like too much, IMO.

smallinsky commented on api/proto/teleport/crownjewel/v1/crownjewel.proto:
```diff
@@ -0,0 +1,90 @@
+}
+
+// AWSTag is a tag that is attached to an AWS resource.
+message AWSTag {
```
nit: Can we reuse https://github.com/gravitational/teleport/blob/994a671a04f8d494120e7f0ee70c229add4e98a4/proto/accessgraph/v1alpha/aws.proto#L169  ?

jakule replied:
Nope. We would need to include our internal proto definition in the API package.

jakule committed:
Update import and label reference in teleport's crownjewel.proto

This commit replaces the TeleportLabel message defined within the crownjewel.proto with an import from label.proto. It refers to labels under the namespace teleport.label.v1, improving modularity and reducing code duplication.
